src/librustc/ty/print/pretty.rs

```diff
@@ -932,90 +960,158 @@ pub trait PrettyPrinter<'tcx>:
                     p!(write("{}{}", data, ui_str))
                 };
             }
-            (ConstValue::Scalar(Scalar::Raw { data, .. }), ty::Int(i)) => {
-                let bit_size = Integer::from_attr(&self.tcx(), SignedInt(*i)).size().bits() as u128;
+            (Scalar::Raw { data, .. }, ty::Int(i)) => {
+                let size = Integer::from_attr(&self.tcx(), SignedInt(*i)).size();
+                let bit_size = size.bits() as u128;
                 let min = 1u128 << (bit_size - 1);
                 let max = min - 1;
 
-                let ty = self.tcx().lift(&ty).unwrap();
-                let size = self.tcx().layout_of(ty::ParamEnv::empty().and(ty)).unwrap().size;
                 let i_str = i.name_str();
                 match data {
                     d if d == min => p!(write("std::{}::MIN", i_str)),
                     d if d == max => p!(write("std::{}::MAX", i_str)),
                     _ => p!(write("{}{}", sign_extend(data, size) as i128, i_str)),
                 }
             }
-            (ConstValue::Scalar(Scalar::Raw { data, .. }), ty::Char) => {
-                p!(write("{:?}", ::std::char::from_u32(data as u32).unwrap()))
+            // Char
+            (Scalar::Raw { data, .. }, ty::Char) => match ::std::char::from_u32(data as u32) {
+                Some(c) => p!(write("{:?}", c)),
+                None => p!(write("{}_char", data)),
+            },
+            // References and pointers
+            (Scalar::Raw { data: 0, .. }, ty::RawPtr(_)) => p!(write("{{null pointer}}")),
+            // This is UB, but we still print it
+            (Scalar::Raw { data: 0, .. }, ty::Ref(_, ty, _)) => {
+                p!(write("{{null reference to "), print(ty), write("}}"))
             }
-            (ConstValue::Scalar(_), ty::RawPtr(_)) => p!(write("{{pointer}}")),
-            (ConstValue::Scalar(Scalar::Ptr(ptr)), ty::FnPtr(_)) => {
+            (Scalar::Raw { data, .. }, ty::Ref(..)) | (Scalar::Raw { data, .. }, ty::RawPtr(_)) => {
+                p!(write("0x{:x}", data))
+            }
```

Comment:
I would replace all of these by printing `{0x0123 as *const T}` and perhaps `{&*0x0123: &T}` (or `{&*(0x0123 as *const T)}`) for references, although the latter is probably also fine to be relegated to the `transmute` case.

Reply:
> replace all of these by printing {0x0123 as *const T}

done

> perhaps {&*0x0123: &T} (or {&*(0x0123 as *const T)}) for references, although the latter is probably also fine to be relegated to the transmute case.

done (relegated to transmute)